File: CodingTemplates/PHP/Model/CommonFunctions.php
<?php
/**
 * Code common to one or more files.
 *
 * PHP version used: 7.3.12
 * SQLite version used: 3.28.0
 *
 * Styling guide: PSR-12: Extended Coding Style
 *     (https://www.php-fig.org/psr/psr-12/)
 *
 * @category  Basics
 * @package   PHP
 * @author    Rob Garcia <rgarcia@rgprogramming.com>
 * @license   https://opensource.org/licenses/MIT The MIT License
 * @link      https://github.com/garciart/Basics
 * @copyright Copyright 2019-2020, Rob Garcia
 */

/**
 * Get the application's model directory.
*/
const MODEL_DIR = __DIR__;

// To run these functions, switch to the model directory.
chdir(MODEL_DIR);

// Report all errors and log them in ErrorLog.txt
error_reporting(-1);
ini_set('log_errors', 1);
ini_set('error_log', MODEL_DIR . DIRECTORY_SEPARATOR . 'ErrorLog.txt');

/*
 * IMPORTANT!
 *
 * FOR DEVELOPMENT ERROR REPORTING:
 * Uncomment ini_set('DISPLAY_ERRORS', 1) and comment out set_error_handler()
 * and set_exception_handler()
 *
 * FOR PRODUCTION ERROR REPORTING:
 * Uncomment set_error_handler() and set_exception_handler() and comment out
 * ini_set('DISPLAY_ERRORS', 1)
 */

// Development error reporting
// ini_set('DISPLAY_ERRORS', 1);

/*
 * Production error reporting
 * Use '32767' instead of 'E_ALL' and make sure to set 'DISPLAY_ERRORS = On'
 * in php.ini
 */
set_error_handler("errorHandler", 32767);
set_exception_handler("exceptionHandler");

/**
 * Error handler. Can be used to redirect users to error page.
 *
 * @param integer $errno   The error report level.
 * @param string  $errstr  The error message.
 * @param string  $errfile The filename with the error.
 * @param integer $errline The line number of the error.
 *
 * @return void
 */
function errorHandler($errno, $errstr, $errfile, $errline)
{
    $error = "Type {$errno} Error: {$errstr} in {$errfile} at line {$errline}.";
    echo "{$error}\n";
    error_log($error);
    // Do not die. Redirect the user to an appropriate error page.
}

/**
 * Exception handler. Can be used to redirect users to exception page.
 *
 * @param string $exception Exception class object.
 *
 * @return void
 */
function exceptionHandler($exception)
{
    echo "Type {$exception->getCode()} Exception: {$exception->getMessage()} " .
        "in {$exception->getFile()} at line {$exception->getLine()}.\n";
    error_log($exception);
    // Do not die. Redirect the user to an appropriate exception page.
}

/**
 * Validate UserID.
 *
 * @param int $userID The UserID that will be entered in the database.
 *
 * @return boolean True if the UserID is an integer greater than 0, false if not.
 */
function validateUserID($userID)
{
    if (empty($userID) || $userID < 1 || !filter_var($userID, FILTER_VALIDATE_INT)) {
        return false;
    } else {
        return true;
    }
}

/**
 * Validate text input.
 *
 * @param string $text The text that will be entered into the database.
 *
 * @return boolean True if the text is valid, false if not.
 */
function validateText($text)
{
    if (empty(trim($text)) ||
        (!preg_match("/^[A-Za-z0-9\s\-._~:\/?#\[\]@!$&'()*+,;=]*$/", $text))
    ) {
        return false;
    } else {
        return true;
    }
}

/**
 * Validate email address.
 *
 * @param string $email The email address that will be entered into the database.
 *
 * @return boolean True if the email is valid, false if not.
 */
function validateEmail($email)
{
    if (empty(trim($email)) || (!filter_var($email, FILTER_VALIDATE_EMAIL))) {
        return false;
    } else {
        return true;
    }
}

/**
 * Validate date format.
 *
 * @param string $date The date that will be entered into the database.
 *
 * @return boolean True if the date format is valid, false if not.
 */
function validateDate($date)
{
    if (empty(trim($date)) ||
        (!preg_match(
            "/^([0-9]){4}-([0-9]){2}-([0-9]){2} ([0-9]){2}:([0-9]){2}:([0-9]){2}$/",
            $date)
        ) ||
        strlen($date) != 19
    ) {
        return false;
    } else {
        return true;
    }
}

File: CodingTemplates/Python/model/common_functions.py
#!python
# -*- coding: utf-8 -*-

"""
Code common to one or more files.

Python version used: 3.6.8
SQLite version used: 3.21.0

Styling guide: PEP 8 -- Style Guide for Python Code
    (https://www.python.org/dev/peps/pep-0008/) and
    PEP 257 -- Docstring Conventions
    (https://www.python.org/dev/peps/pep-0257/)
"""

from __future__ import print_function

import datetime
import logging
import os
import re
import sys

# Module metadata dunders
__author__ = "Rob Garcia"
__copyright__ = "Copyright 2019-2020, Rob Garcia"
__email__ = "rgarcia@rgprogramming.com"
__license__ = "MIT"


# Used os instead of os and Pathlib (redundant) to set PWD to correct directory

# Get the application's model directory.
MODEL_DIR = os.path.dirname(__file__)

# To run these functions, switch to the model directory.
os.chdir(MODEL_DIR)

# Enable error and exception logging
logging.basicConfig(
    filename="error_log.txt", filemode="a+", level=logging.DEBUG)
# Set to True during development and to False during production
DISPLAY_ERRORS = True


def log_error(exc_info):
    """Reformats error and exception details and records them in plain text in
        the error_log file.

    :param exc_info: Exception details from sys module.
    :type exc_info: tuple
    :return: Reformated exception details in plain text.
    :rtype: str
    """
    exception = None
    try:
        exc_type, exc_value, exc_traceback = exc_info
        local_timezone = datetime.datetime.now(
            datetime.timezone.utc).astimezone().tzinfo
        time_stamp = datetime.datetime.now(
            local_timezone).strftime("%Y-%m-%d %H:%M:%S %Z")
        exception = ("[{}] Type {} Exception: {} in {} at line {}.".format(
            time_stamp,
            exc_type.__name__,
            exc_value,
            exc_traceback.tb_frame.f_code.co_filename,
            exc_traceback.tb_lineno))
        logging.debug(exception)
    except Exception:
        exc = sys.exc_info()
        # if co.DISPLAY_ERRORS:
        print(exc)
    return exception


def validate_user_id(user_id):
    """Validate UserID.

    :param user_id: The user ID that will be entered in the database.
    :type user_id: int
    :return: True if the user ID is an integer greater than 0, false if not.
    :rtype: bool
    """
    if not user_id or user_id <= 0:
        return False
    else:
        return True


def validate_text(text):
    """Validate text input.

    :param text: The text that will be entered into the database.
    :type text: str
    :return: True if the text is valid, false if not.
    :rtype: bool
    """
    if not (text.strip()) or re.search(
            r"^[A-Za-z0-9 -._~:/?#[\]@!$&'()\*+,;=]*$", text.strip()) is None:
        return False
    else:
        return True


def validate_email(email):
    """Validate email address.

    :param email: The email address that will be entered into the database.
    :type email: str
    :return: True if the email is valid, false if not.
    :rtype: bool
    """
    if not (email.strip()) or re.search(
            r"^[A-Za-z0-9-._~/?#!$&'%*+=`{|}^]+@[A-Za-z0-9.-]+$", email) is None:
        return False
    else:
        return True


def validate_date(date):
    """Validate date format.

    :param date: The date that will be entered into the database.
    :type date: str
    :return: True if the date format is valid, false if not.
    :rtype: bool
    """
    regex = re.compile(".{4}-.{2}-.{2} .{2}:.{2}:.{2}")
    date_format = "xxxx-xx-xx xx:xx:xx"
    if not (date.strip()) or regex.match(date_format) is None or len(date_format) != 19:
        return False
    else:
        return True
